lib/connections/limiter.go

```diff
@@ -12,44 +12,161 @@ import (
+	for _, v := range to.Devices {
+		if v.DeviceID == to.MyID {
+			// This limiter was created for local device. Should skip this device
+			continue
+		}
+		seen[v.DeviceID] = struct{}{}
+
+		readLimiter, okR := lim.getDeviceReadLimiter(v.DeviceID)
+		writeLimiter, okW := lim.getDeviceWriteLimiter(v.DeviceID)
+
+		// There was no limiter for this ID in map.
+		// This means that we added this device and we should create new limiter
+		if !okR && !okW {
+			readLimiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+			writeLimiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+
+			lim.setDeviceLimiters(v.DeviceID, writeLimiter, readLimiter)
+		} else if !okR || !okW {
+			// One of the read/write limiters is not present while the
+			// corresponding write/read one exists. Something has gone wrong.
+			panic("broken symmetry in device read/write limiters")
+		}
+
+		// limiters for this device are created so we can store previous rates for logging
+		previousReadLimit := readLimiter.Limit()
+		previousWriteLimit := writeLimiter.Limit()
+
+		// Nothing about this device has changed. Start processing next device
+		if okR && okW &&
+			rate.Limit(v.MaxRecvKbps)*1024 == previousReadLimit &&
+			rate.Limit(v.MaxSendKbps)*1024 == previousWriteLimit {
+			continue
+		}
+		setLimitsForDevice(v, readLimiter, writeLimiter)
+
+		readLimitStr := "is unlimited"
+		if v.MaxRecvKbps > 0 {
+			readLimitStr = fmt.Sprintf("limit is %d KiB/s", v.MaxRecvKbps)
+		}
+		writeLimitStr := "is unlimited"
+		if v.MaxSendKbps > 0 {
+			writeLimitStr = fmt.Sprintf("limit is %d KiB/s", v.MaxSendKbps)
+		}
+
+		l.Infof("Device %s send rate %s, receive rate %s", v.DeviceID, readLimitStr, writeLimitStr)
+	}
+
+	// Delete remote devices which were removed in new configuration
+	for _, v := range from.Devices {
+		if _, ok := seen[v.DeviceID]; !ok {
+			l.Debugf("deviceID: %s should be removed", v.DeviceID)
+			lim.deleteDeviceLimiters(v.DeviceID)
+		}
+	}
+
+	l.Debugln("Rebuild of device limiters map finished")
+}
+
+// Compare read/write limits in configurations
+func deviceLimitsChanged(from, to config.Configuration) bool {
+	if len(from.Devices) != len(to.Devices) {
+		return true
+	}
+	fromMap := config.MapDeviceConfigs(from.Devices)
+	toMap := config.MapDeviceConfigs(to.Devices)
+	// len(from.Devices) == len(to.Devices) so we can do range from.Devices
+	for k := range fromMap {
+		if _, ok := toMap[k]; !ok {
+			return true
+		}
+		// Read/write limits were changed for this device
+		if fromMap[k].MaxSendKbps != toMap[k].MaxSendKbps ||
+			fromMap[k].MaxRecvKbps != toMap[k].MaxRecvKbps {
+			return true
+		}
+	}
+	return false
 }
 
 func (lim *limiter) VerifyConfiguration(from, to config.Configuration) error {
 	return nil
 }
 
 func (lim *limiter) CommitConfiguration(from, to config.Configuration) bool {
+	// to ensure atomic update of configuration
+	lim.mu.Lock()
+	defer lim.mu.Unlock()
+	if deviceLimitsChanged(from, to) {
```

Comment:
I would remove this check. You already do the same checks in `processDevicesConfiguration` and skip any devices which don't change.